packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -31,19 +33,160 @@ import PropTypes from 'prop-types'
  *
  */
 export const CategoriesContext = React.createContext()
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children, testData = {}}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: false
+        }))
+    })
+    const [queue, setQueue] = useState({})
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+
+    // copied from https://github.com/brojd/obj-traverse/blob/master/src/findAndModifyFirst/findAndModifyFirst.js
```

Comment:
That code is MIT licensed. Are we allowed to take it? At minimum, we'll need to credit somewhere.

Reply:
Also, it has some patterns that should be cleaned up like assignment within ternaries, unguarded `for...in`, etc.

Reply:
This isn't art. We can remove the comment citing the original if people prefer, there are about 1 million instances of people writing recursive tree traversal code.

Reply:
I'd be sympathetic to the position if it weren't something that was specifically authored countless times already. This is not even a little bit unique though

Reply:
Removing the comment that cites the reference